src/Generating/FinishGen.cpp

```diff
@@ -665,51 +665,106 @@ void cFinishGenVines::GenFinish(cChunkDesc & a_ChunkDesc)
 
 
 
-
-
 ////////////////////////////////////////////////////////////////////////////////
 // cFinishGenSprinkleFoliage:
 
-bool cFinishGenSprinkleFoliage::TryAddSugarcane(cChunkDesc & a_ChunkDesc, int a_RelX, int a_RelY, int a_RelZ)
+bool cFinishGenSprinkleFoliage::TryAddCactus(cChunkDesc &a_ChunkDesc, int a_RelX, HEIGHTTYPE & a_RelY, int a_RelZ)
 {
-	// We'll be doing comparison to neighbors, so require the coords to be 1 block away from the chunk edges:
-	if (
-		(a_RelX < 1) || (a_RelX >= cChunkDef::Width  - 1) ||
-		(a_RelY < 1) || (a_RelY >= cChunkDef::Height - 2) ||
-		(a_RelZ < 1) || (a_RelZ >= cChunkDef::Width  - 1)
-	)
+	if (!IsDesertVariant(a_ChunkDesc.GetBiome(a_RelX, a_RelZ)))
 	{
 		return false;
 	}
 
-	// Only allow dirt, grass or sand below sugarcane:
-	switch (a_ChunkDesc.GetBlockType(a_RelX, a_RelY, a_RelZ))
+	for (int i = 0; i < 1 + (m_Noise.IntNoise2DInt(a_RelX, a_RelZ) % m_MaxCactusHeight); i++)
 	{
-		case E_BLOCK_DIRT:
-		case E_BLOCK_GRASS:
-		case E_BLOCK_SAND:
+		if (
+			(a_RelX < 1) || (a_RelX >= cChunkDef::Width - 1) ||
+			(a_RelY < 1) || (a_RelY >= cChunkDef::Height - 2) ||
+			(a_RelZ < 1) || (a_RelZ >= cChunkDef::Width - 1)
+		)
 		{
-			break;
+			return (i == 0) ? false : true;
 		}
-		default:
+
+		switch (a_ChunkDesc.GetBlockType(a_RelX, a_RelY, a_RelZ))
 		{
-			return false;
+			case E_BLOCK_SAND :
+			case E_BLOCK_CACTUS :
```

Comment:
You forgot red sand 😏 .

Reply:
For that to make a difference, the switch in `GenFinish` would need updating as well.

Reply:
Isn't red sand a meta type of the `E_BLOCK_SAND` (l. 767 in `BlockID.h`) ? Using `E_BLOCK_SAND` does not handles both ?

Reply:
[You're right](https://minecraft.gamepedia.com/Sand)

Reply:
Oh men. Must have been the day, I've learned about the meta removal in 1.13.